## 자료구조

------------------

### \# 기본개념

1. 자료구조의 효율성은 수행되는 연산의 `수행시간`으로 측정
    - `수행시간` 측정 = `시간복잡도` * 공간복잡도



2. 대부분의 알고리즘들이 비슷한 크기의 메모리 공간을 사용하기 때문에 `시간복잡도`가 중점이다.
    - `시간복잡도`는 수행시간으로 명시하기도 한다.

**[!TIP]** 한줄요약: 수행시간은 시간복잡도와 공간복잡도를 기반하여 분석하지만 대부분 알고리즘은 비슷한 메모리 공간을 차지하여 시간복잡도(수행시간)이 중요하다.

### \# 표기법은 총 3가지가 있다

#### Big-Oh 표기법
- 이보다 너 나쁠 수는 없다.(점근적 상한)
    - 최악의 수행 시간이 될 수 있는 가능성을 판단

#### Big-Omega 표기법
- 이보다 더 좋을 수는 없다.(점근적 하한)

#### Theta 표기법
- Big-Oh 와 Big-Omega 동시 성립



**[!TIP]** 점근적: 점점 가까워짐

### <p style="text-align: center;">[알고리즘의 수행시간]</p>

| 표기     | 시간           |          |
|----------|----------------|----------|
| O(1)     | 상수시간       | 효율적   |
| O(logN)  | 로그(대수)시간 |          |
| O(N)     | 선형시간       |          |
| O(NlogN) | 로그선형시간   |          |
| O(N^2)   | 제곱시간       |          |
| O(N^3)   | 세제곱시간     |          |
| O(2^N)   | 지수시간       | 비효율적 |

<p style="text-align: center;">*(주로 O-표기를 사용)</p>

-------------------

### \# 파이썬 언어 기본지식

In [1]:
a = [] # 비어있는 리스트 a 선언
b = [None] * 10 # 크기가 10//각 원소가 None 로 초기화 된 리스트
c = [40, 10, 70, 60] # 크기가 4이고, 4개의 정수로 초기화 된 리스트

**[!TIP]**
- 리스트에 들어있는 값을 개수로 하지 않고 `크기`라고 부름
    - 배열은 동일한 타입의 항목들을 저장
    - 리스트는 다른 타입의 항목들을 저장

### \# 의사난수(Pseudo-random Number) 생성

In [2]:
import random
import time
random.seed(time.time())
a = []
for i in range(100):
    a.append(random.randint(1, 1000))
    
start_time = time.time()


# 같은 포맷팅 방식인데 결과값이 달라보인다.
# new 포맷팅은 e-05 에 주목: 5.62.... 에 ^-5 (마이너스 5승) 으로 보면 된다.
print('--- {} seconds ---'.format((time.time() - start_time)))
print('--- %s seconds ---' % (time.time() - start_time))

--- 5.459785461425781e-05 seconds ---
--- 0.001172780990600586 seconds ---


In [3]:
# 포맷팅 문서를 보니 포맷팅으로 인덱싱, 슬라이싱이 가능한 것을 확인.
# 언젠가는 사용할 수도 있을 것 같아 기억만 해두자. (*자세한 내용은 관련 문서 참조)

'{:>10}'.format('test')

'      test'

### \# 내장함수

In [4]:
a = [1, 5, 4, 2, 7, 10, 11, 12]

**lambda 함수를 사용할 경우**

```
- lambda 인자(arguments): 식(expression)
```

In [5]:
even = list(filter(lambda x: (x%2 == 0), a))
print(even) # type: list

[4, 2, 10, 12]


In [6]:
ten_times = list(map(lambda x: x * 10, a))
print(ten_times) # list a 의 각 숫자 * 10 를 리턴

[10, 50, 40, 20, 70, 100, 110, 120]


In [7]:
b = [[0, 1, 8], [7, 2, 2], [5, 3, 10], [1, 4, 5]]

In [8]:
b.sort(key = lambda x: x[2])
print(b) # list 의 마지막 숫자를 기준으로 정렬 b[[2번째], [4번째], [1번째], [3번째]]

[[7, 2, 2], [1, 4, 5], [0, 1, 8], [5, 3, 10]]


**함수로 사용할 경우**

In [9]:
def even_method(a):
    result = []
    for i in a:
        if i % 2 == 0:
            result.append(i)
#     return result
    print(type(result))
    print(result)

In [10]:
even_method(a)

<class 'list'>
[4, 2, 10, 12]


**리스트 컴프리헨션으로 구현**

In [11]:
[i for i in a if i % 2 == 0] # type: list

[4, 2, 10, 12]

### \# 단순연결리스트

- 단순연결리스트는 `동적 메모리 할당`을 이용해 노드들을 한 방향으로 연결하여 리스트를 구현
    - 동적 메모리 할당: 실행 시간동안에 메모리를 사용(C 언어는 가비지 컬렉터가 없어서 직접 해제해야 하지만 파이썬은 자동 해제)

In [12]:
class SList:
    class Node:
        def __init__(self, item, link):
            self.item = item
            self.next = link
            
    def __init__(self):
        self.head = None
        self.size = 0

    def size(self):
        return self.size

    def is_empty(self):
        return self.size == 0

    def insert_front(self, item):  # 첫 노드에 삽입
        # 첫번째 노드가 비어있으면
        if self.is_empty():
            # 새로 만든 Node 객체를 self.head 에 참조
            self.head = self.Node(item, None)
        else:
            # Node 에 데이터가 있다면 이번에 들어오는 Node 값을 head 에 참조
            self.head = self.Node(item, self.head)
        # if, else 문에서 Node 객체가 삽입되었다면 size + 1
        self.size += 1

    def insert_after(self, item, p):    # 지정한 pointer 노드 다음에 삽입
        # pointer.next 에 노드 삽입 = Node 객체 생성(pointer 노드 다음에 item 을 삽입)
        p.next = SList.Node(item, p.next)
        self.size += 1

    def delete_front(self):    # 첫 번째 노드를 삭제할 때
        # 만약 지정한 노드 객체가 없다면
        if self.is_empty():
            # 에러를 발생
            raise EmptyError('Underflow')
        # 지정한 노드(item) 이 있고, 그게 head 라면
        else:
            # self.head 의 다음 노드를 self.head 로 참조
            self.head = self.head.next
            self.size -= 1

    def delete_after(self, p):
        if self.is_empty():
            raise EmptyError('Under Flow')
        else:
            # p.next 가 t 와 동일한 3번째 노드에 있다. 그런데 t 가 4번째 노드로 가서 p.next 를 선언
            # 이렇게 되면 기존의 p.next 는 노드의 연결이 끊기게 됨
            # 한줄요약: t 의 배신

            # 현재 t 와 p.next 는 동일한 곳을 가리키고 있는 상태
            t = p.next
            # t 는 p.next 의 다음 노드를 가리키고 p.next 에 할당
            p.next = t.next
            self.size -= 1

    def search(self, target):
        p = self.head
        # 노드가 5개라면 총 5번을 순회(k: 인덱스)
        for k in range(self.size):
            # 파라미터로 받은 target 과 p.item 을 비교하여
            if target == p.item:
                # 인덱스 값을 반환
                return k
            # 만약 target 과 p.item 이 다르다면 p 를 다음 노드로 이동시킴
            p = p.next
        # 검색에 실패할 경우 None 를 반환
        return None

    def print_list(self):
        p = self.head
        while p:
            if p != None:
                print(p.item, '->', end='')
            else:
                print(p.item)
            p = p.next

class EmptyError(Exception):
    pass


In [13]:
s = SList()
s.insert_front('orange')
s.insert_front('apple')
s.insert_after('cherry', s.head.next)
s.insert_front('pear')
s.print_list()
print('cherry 는 {}번째'.format(s.search('cherry')))
print('kiwi 는'.format(s.search('kiwi')))
s.delete_after(s.head)
s.print_list()
print('첫 노드 삭제 후:\t \t', end='')
s.delete_front()
s.print_list()
print('첫 노드로 망고, 딸기 삽입 후:\t', end='')
s.insert_front('mango')
s.insert_front('strawberry')
s.print_list()
s.delete_after(s.head.next.next)
print('오렌지 다음 노드 삭제 후:\t', end='')
s.print_list()

pear ->apple ->orange ->cherry ->cherry 는 3번째
kiwi 는
pear ->orange ->cherry ->첫 노드 삭제 후:	 	orange ->cherry ->첫 노드로 망고, 딸기 삽입 후:	strawberry ->mango ->orange ->cherry ->오렌지 다음 노드 삭제 후:	strawberry ->mango ->orange ->

### \# 이중연결리스트

이중연결리스트는 각 노드가 두 개의 레퍼런스를 가지고 각각 이전 노드와 다음 노드를 가리키는 연결리스트

In [14]:
class DList:
    class Node:
        def __init__(self, item, prev, link):
            self.item = item
            self.prev = prev
            self.next = link

    def __init__(self):
        self.head = self.Node(None, None, None)
        self.tail = self.Node(None, self.head, None)
        self.head.next = self.tail
        self.size = 0

    def size(self):
        return self.size

    def is_empty(self):
        return self.size == 0

    def insert_before(self, p, item):   # 지정한 p 노드 앞에 삽입
        # t 변수에 지정한 pointer 의 앞 객체를 참조
        t = p.prev
        # Node 객체를 만드는데 prev 는 p.prev, p 는 새로운 Node 가 아닌 '뒤로 밀리게 되는 Node'가 된다
        # 앞과 뒤의 Node 를 연결하고 n 변수에 할당
        n = self.Node(item, t, p)
        # 새로 만들어진 Node 객체 n 의 앞과 뒤를 연결해준다
        p.prev = n
        t.next = n
        # Node 객체가 만들어졌기 때문에 데이터의 크기(size +1) 을 시켜줌
        self.size += 1

    def insert_after(self, p, item):    # 지정한 p 노드 뒤에 삽입
        # t 변수에 지정한 pointer 의 다음 객체를 참조
        t = p.next
        # Node 객체를 만드는데 p 에 현재 지정한 Node, t 에는 현재 지정한 객체의 다음 Node 를 가리킴
        # p 와 t 노드 사이에 n Node 객체를 만들어줌
        n = self.Node(item, p, t)
        # 새로 만들어진 n Node 객체의 앞과 뒤(p와 t)를 n 에 연결시켜준다
        t.prev = n
        p.next = n

    def delete(self, x):    # x 의 연결을 끊고, f 와 r 을 이어줌
        # forward 변수에 x 의 앞 노드를 가리킴
        f = x.prev
        # rear 변수에 x 의 다음 노드를 가리킴
        r = x.next
        # x 노드를 삭제하기 위해 f 노드는 을 가리키고, r 노드는 f 를 가리킴
        # 이로써 x 의 연결이 끊어짐, 연결이 끊어졌다는건 이 연결 리스트에서 제외되었다는 뜻
        f.next = r
        r.prev = f
        self.size -= 1
        # x 의 실제 값을 return 시켜줌
        return x.item

    def print_list(self):
        if self.is_empty():
            print('리스트 비어있음')
        else:
            # pointer 가 self.head 가 아닌 self.head.next 를 가리키는 이유는??
            # 내 생각에는 pointer 가 self.head 를 가리켜야 head 부터 p.item 을 print 할텐데..
            p = self.head.next
            # p 가 tail 이 아니라면 while 문 안에서 무한 순회
            while p != self.tail:
                # pointer 의 next Node 객체가 tail 이 아니라면
                if p.next != self.tail:
                    # p.item(입력한 값)을 출력
                    print(p.item, ' <--> ', end='')
                # self.tail 일 경우에는
                else:
                    # p.item(입력한 값)을 출력, 이 값은 tail 의 값이 되어 while 문 종료
                    print(p.item)
                # 위의 if 문이 true 이면 pointer 는 다음 Node 객체를 가리키게 설정
                p = p.next


class EmptyError(Exception):
    pass


In [15]:
s = DList()
s.insert_after(s.head, 'apple')
s.insert_before(s.tail, 'orange')
s.insert_before(s.tail, 'cherry')
s.insert_after(s.head.next, 'pear')
s.print_list()
print('마지막 노드 삭제 후:\t', end='')
s.delete(s.tail.prev)
s.print_list()
print('맨 끝에 포도 삽입 후:\t', end='')
s.insert_before(s.tail, 'grape')
s.print_list()
print('첫 노드 삭제 후:\t', end='')
s.delete(s.head.next)
s.print_list()
print('첫 노드 삭제 후:\t', end='')
s.delete(s.head.next)
s.print_list()
print('첫 노드 삭제 후:\t', end='')
s.delete(s.head.next)
s.print_list()

apple  <--> pear  <--> orange  <--> cherry
마지막 노드 삭제 후:	apple  <--> pear  <--> orange
맨 끝에 포도 삽입 후:	apple  <--> pear  <--> orange  <--> grape
첫 노드 삭제 후:	pear  <--> orange  <--> grape
첫 노드 삭제 후:	리스트 비어있음
첫 노드 삭제 후:	grape


### \# 원형연결리스트

In [16]:
class CList:
    class _Node:
        def __init__(self, item, link):
            self.item = item
            self.next = link

    def __init__(self):
        self.last = None
        self.size = 0

    # Node 의 size(데이터 크기)를 리턴
    def no_items(self):
        return self.size

    # Node 가 비어있을 때의 상황을 함수화 시켜놓음
    # self.is_empty() 로 메서드를 불러 사용하기 때문에 가독성이 좋아짐
    def is_empty(self):
        return self.size == 0

    def insert(self, item):
        # n 변수에 새로운 Node 객체를 생성
        n = self._Node(item, None)
        # Note List 가 비어있다면
        if self.is_empty():
            # Node 객체의 next 를 자신을 참조하면서, last 로 참조(레퍼런스 카운트 2)
            n.next = n
            self.last = n
        # CList 에 Node 객체가 존재한다면
        else:
            # self.last.next(즉 첫번째 생성한 Node 의 다음) 에 n.next 할당
            n.next = self.last.next
            # 새로 만든 Node 는 첫번째 노드의 다음에 위치한다
            self.last.next = n
        self.size += 1

    def first(self):
        if self.is_empty():
            raise EmptyError('Underflow')
        f = self.last.next
        return f.item

    def delete(self):
        if self.is_empty():
            raise EmptyError('Underflow')
        x = self.last.next
        if self.size == 1:
            self.last = None
        else:
            self.last.next = x.next
        self.size -= 1
        return x.item

    def print_list(self):
        if self.is_empty():
            print('리스트 비어있음')
        else:
            f = self.last.next
            p = f
            while p.next != f:
                print(p.item, ' -> ', end='')
                p = p.next
            print(p.item)


class EmptyError(Exception):
    pass

In [17]:
s = CList()
s.insert('pear')
s.insert('cherry')
s.insert('orange')
s.insert('apple')
s.print_list()
print('s의 길이 =', s.no_items())
print('s의 첫 항목 :', s.first())
s.delete()
print('첫번째 노드 삭제 후: ', end='')
s.print_list()
print('s의 길이 =', s.no_items())
print('s의 첫 항목: ', s.first())
s.delete()
print('첫번째 노드 삭제 후: ', end='')
s.print_list()
s.delete()
print('첫번째 노드 삭제 후: ', end='')
s.print_list()
s.delete()
print('첫번째 노드 삭제 후: ', end='')
s.print_list()

apple  -> orange  -> cherry  -> pear
s의 길이 = 4
s의 첫 항목 : apple
첫번째 노드 삭제 후: orange  -> cherry  -> pear
s의 길이 = 3
s의 첫 항목:  orange
첫번째 노드 삭제 후: cherry  -> pear
첫번째 노드 삭제 후: pear
첫번째 노드 삭제 후: 리스트 비어있음


### \# 스택과 큐

In [18]:
class Node:
    # 전역변수 top 에 생성자함수를 사용한 노드를 생성
    # top = Node(item, top)
    def __init__(self, item, link):
        self.item = item
        self.next = link
        
def push(item):
    # 전역변수 top/size 선언
    global top
    global size
    # 새로운 노드를 push 할 때는 스택구조이기 때문에, 항상 top = 새로 생성한 노드
    top = Node(item, top)
    size += 1
    
def peek():
    if size != 0:
        # top 인스턴스의 item 속성에 접근
        return top.item
    
def pop():
    # pop 은 해당 스택에서 최상위에 있는 리스트의 값을 꺼내야 함
    global top
    global size
    if size != 0:
        # 해당 스택에서 top.item 을 임의의 변수 top_item 으로 할당
        # (이 변수에 할당된 값이 return 됨)
        top_item = top.item
        # 바로 아래에 있는 리스트의 값을 새로운 top 으로 포인터 지정
        top = top.next
        size -= 1
        # 위에서 지정한 임의의 변수 top_item 을 리턴
        # 리턴은 함수를 부르면 1번만 실행되기 때문에
        # top_item 이 존재하지 않는다는 에러메시지는 발생하지 않음
        return top_item
    
def print_stack():
    print('top -> ', end='')
    p = top
    while p:
        # stack 구조상 바로 아래에 아이템이 있을 경우 if 처리
        if p.next != None:
            print(p.item, '\n', end='')
        # stack 에서 가장 아래에 아이템일 경우 else 처리
        else:
            print(p.item, '\n', end='')
        # if, else 문에서 처리하고 바로 아래의 item 을 .next 에 할당
        p = p.next
    print()


top = None
size = 0
push('apple')
push('orange')
push('cherry')
print('사과, 오렌지, 체리  push 후\t', end='')
print()

print_stack()
print('top 항목:', end='')
print(peek())
print()

push('pear')
print('[배(pear) push 후]\n', end='')
print_stack()
pop()
push('grape')
print('[pop(), 포도(grape) push 후]\n', end='')
print_stack()

사과, 오렌지, 체리  push 후	
top -> cherry 
orange 
apple 

top 항목:cherry

[배(pear) push 후]
top -> pear 
cherry 
orange 
apple 

[pop(), 포도(grape) push 후]
top -> grape 
cherry 
orange 
apple 



[!TIP]
> 알고리즘의 수행시간은 주로 big-O 표기법(이보다 나쁠 수는 없다.) 로 표현하며, 단순연결리스트로 구현한 stack 에서 push, pop 은 `O(1)` 시간이 소요

> 그러나 우리 눈에는 보이지 않지만 파이썬의 리스트 크기는 동적으로 자동으로 확장/축소 되며, 모든 항목은 새로운 리스트로 복사해야 하기 때문에 `O(N)` 시간이 소요

| 표기     | 시간           | 비고         |
|----------|----------------|----------|
| O(1)     | 상수시간       | 효율적 **(단순연결리스트 push, pop)**   |
| O(logN)  | 로그(대수)시간 |          |
| O(N)     | 선형시간       |          |
| O(NlogN) | 로그선형시간   |          |
| O(N^2)   | 제곱시간       |          |
| O(N^3)   | 세제곱시간     |          |
| O(2^N)   | 지수시간       | 비효율적 |

<p style="text-align: center;">*(주로 O-표기를 사용)</p>

#### \# 스택의 응용

1. 괄호 짝 맞추기
    - 왼쪽 괄호는 스택에 push, 오른쪽 괄호는 스택에 pop 하여 짝이 맞지 않을 경우 에러를 발생


2. 회문(Palindrome)
    - 앞으로부터 읽으나 뒤로부터 읽으나 같은 string type 을 검사  예) RACECAR
    - 전반부의 문자들을 스택에 push, 후반부의 각 문자를 차례로 pop 한 문자와 비교



\# RACECAR

\# RAC 는 stack 에 push 하고, pop은 역순으로 나오기 때문에 CAR 이 나옴

step7. pop(`R`)

step6. pop(`A`)

step5. pop(`C`)

step4. `E` 는 읽고 버림

step3. push(`C`)

step2. push(`A`)

step1. push(`R`)

### \# 큐

- 단순연결리스트로 구현한 큐
    - CPU 의 태스트 스케줄링, 네트워크 프린터, 실시간 시스템의 인터럽트 처리, 다양한 이벤트 구동 방식 컴퓨터 시뮬레이션, 콜 센터의 전화 서비스 처리 등
    - 이진트리의 레벨순회, 그래프의 너비우선탐색

In [19]:
class Node:
    def __init__(self, item, n):
        self.item = item
        self.next = n
        
def add(item):
    # 전역변수로 size, front, rear 선언
    # 전역변수를 선언하면 해당 함수가 종료되도 메모리 공간 어디엔가 존재
    global size
    global front
    global rear
    # 새로운 Node 를 생성하고, link 는 없음
    new_node = Node(item, None)
    if size == 0:
        front = new_node
    else:
        # 이미 Node 가 존재한다면 리스트의 맨 뒤.next 를 새로 만든 Node 로 지정
        rear.next = new_node
    rear = new_node
    size += 1

# remove 는 파라미터를 따로 받지 않고, 전역변수 자체에 접근하여 연산 처리를 진행
def remove():
    global size
    global front
    global rear
    if size != 0:
        # 큐 구조는 맨 처음에 들어간 front 의 값을 return 시킴
        # front.item 을 따로 변수로 지정, front.next 의 값을 front 변수로 지정
        fitem = front.item
        front = front.next
        size -= 1
        if size == 0:
            rear = None
        # return 시킴으로 인해서 리스트의 길이가 -1 됨
        return fitem
    
def print_q():
    p = front
    print('front: ', end='')
    while p:
        if p.next != None:
            print(p.item, '-> ', end='')
        else:
            print(p.item, end='')
        p = p.next
    print(' : rear')

In [20]:
front = None
rear = None
size = 0

add('apple')
add('orange')
add('cherry')
add('pear')
print('사과, 오렌지, 체리, 배 삽입 후: \t', end='')
print_q()
remove()
print('remove한 후:\t\t', end='')
print_q()
remove()
print('remove한 후:\t\t', end='')
print_q()
add('grape')
print('포도 삽입 후:\t\t', end='')
print_q()

사과, 오렌지, 체리, 배 삽입 후: 	front: apple -> orange -> cherry -> pear : rear
remove한 후:		front: orange -> cherry -> pear : rear
remove한 후:		front: cherry -> pear : rear
포도 삽입 후:		front: cherry -> pear -> grape : rear


### \# 데크(Deque)

- (Double-ended Queue, Deque)는 양쪽 끝에서 삽입과 삭제를 허용하는 자료구조
- 데크는 스택과 큐 자료구조를 혼합한 자료구조


In [21]:
from collections import deque
dq = deque('data') # deque 객체 생성, deque(['d', 'a', 't', 'a'])
for elem in dq:
    print(elem.upper())
print()
# 하나의 문자열로 해당 리스트에 삽입
dq.append('r')
dq.appendleft('k')
print(dq)
dq.pop()
dq.popleft()
print(dq[-1])
print('x' in dq)
# 문자열이 아닌 각 문자마다 dq 리스트에 추가
dq.extend('structure')
dq.extendleft(reversed('python'))
print(dq)

D
A
T
A

deque(['k', 'd', 'a', 't', 'a', 'r'])
a
False
deque(['p', 'y', 't', 'h', 'o', 'n', 'd', 'a', 't', 'a', 's', 't', 'r', 'u', 'c', 't', 'u', 'r', 'e'])


### \# 트리

#### 이진트리

1. 각 노드의 자식 수가 2 이하인 트리
    - 이진트리는 1. 비어있거나 // 2. 비어있지 않다면 루트와 2개의 이진트리인 왼쪽 서브트리와 오른쪽 서브트리로 구성
    

2. 이진트리의 형태
    - 포화이진트리, 완전이진트리(또는 불완전한 이진트리)
    
    
3. 리스트에 저장해야 효율적인 이진트리와 그렇지 않은 이진트리
    - 완전이진트리(또는 포화이진트리)는 리스트에 꽉꽉 채워져있어(메모리에 순서대로 저장) 효율적이다.
    - 그렇지만 편향이진트리는 리스트에 듬성듬성 값이 채워져있어(메모리에 듬성듬성 저장) 비효율적이다.


4. 재귀함수로 호출
    - 재귀함수로 호출하면 바로 이전의 데이터값은 재귀함수의 호출 위치에 머물러있다가, 재귀함수 호출값이 다음 코드로 넘어갔을 때 그 함수 호출 값을 처리한다......(말로 하니 어렵다)
    
    
5. 스택프레임에 쌓이는 것은 재귀함수
    - 레벨순회를 제외하고는 모두 스택 자료구조를 사용
    - 함수의 재귀호출은 시스템 스택을 사용하므로 스택 자료구조를 사용한 것으로 간주
    - 스택에 사용되는 메모리 공간의 크기는 트리의 높이에 비례

In [22]:
class Node:
    def __init__(self, item, left=None, right=None):
        self.item = item
        self.left = left
        self.right = right
        
class BinaryTree:
    def __init__(self):
        self.root = None
        
    def preorder(self, n):
        # 재귀함수로 root 를 기준으로 계속해서 하위 레벨의 노드를 호출
        # 최하위의 레벨의 노드가 종료되면, 그 부모 레벨의 노드에 대한 print()
        # 즉, 1레벨, 2레벨, 3레벨의 순서로 스택프레임은 쌓였기 때문에 3레벨, 2레벨, 1레벨의 순으로 연산 처리를 진행
        if n != None:
            print(str(n.item), ' ', end='')
            if n.left:
                self.preorder(n.left)
            if n.right:
                self.preorder(n.right)
                
    def inorder(self, n):
        if n != None:
            if n.left:
                self.preorder(n.left) # 루트 노드부터 스택프레임이 차곡차곡 쌓여있음
            print(str(n.item), ' ', end='')
            if n.right:
                self.preorder(n.right)
                
    def postorder(self, n):
        if n != None:
            if n.left:
                self.postorder(n.left)
            if n.right:
                self.postorder(n.right)
            print(str(n.item), ' ', end='')
            
    def levelorder(self, root):
        q = []
        q.append(root)
        # 노드는 while 문을 기준으로 계속 순회한다
        while len(q) != 0:
            t = q.pop(0)
            print(str(t.item), ' ', end='')
            # 부모노드의 자식노드들을 계속해서 q 리스트에 추가, 그러나 print()는 부모 노드가 출력
            if t.left:
                q.append(t.left)
            if t.right:
                q.append(t.right)

    def height(self, root):
        if root == None:
            return 0
        return max(self.height(root.left), self.height(root.right)) + 1
    

In [23]:
t = BinaryTree()
n1 = Node(100)
n2 = Node(200)
n3 = Node(300)
n4 = Node(400)
n5 = Node(500)
n6 = Node(600)
n7 = Node(700)
n8 = Node(800)
n1.left = n2
n1.right = n3
n2.left = n4
n2.right = n5
n3.left = n6
n3.right = n7
n4.left = n8
t.root = n1

print(' 트리 높이 =', t.height(t.root))
print(' 전위순회: \t', end='')
t.preorder(t.root)
print('\n 중위순회: \t', end='')
t.inorder(t.root)
print('\n 후위순회: \t', end='')
t.postorder(t.root)
print('\n 레벨순회: \t', end='')
t.levelorder(t.root)

 트리 높이 = 4
 전위순회: 	100  200  400  800  500  300  600  700  
 중위순회: 	200  400  800  500  100  300  600  700  
 후위순회: 	800  400  500  200  600  700  300  100  
 레벨순회: 	100  200  300  400  500  600  700  800  

### \# 이진힙

1. 최소힙과 최대힙으로 구분
    - 최소힙에서 최소값의 삭제는 루트 노드를 삭제하고, 최하위(최대값)을 루트노드에 삽입하여 자식 노드와 비교하여 결과값 리턴(다운힙)
    
    
2. 이진힙의 사용
    - 우선순위를 가진 데이터를 처리하는 자료구조로서, 관공서, 은행, 병원, 우체국, 대형마켓, 공항 등에서 이루어지는 업무와 관련된 이벤트 처리, 컴퓨터 운영체제의 프로세스 처리, 네트워크 라우터에서의 패킷 처리 등에 적합한 자료구조
    - **실시간 급상승 검색어(데이터 스트림에서 Top k 항목을 유지) 제공을 위한 적절한 자료구조**
    
    
3. 검색어 구현
    - real_time_rank = \[[76, '파이썬'], [43, '자바'], [32, '고']\]
        - 여기서 갑자기 \[[120, '인공지능']\] 검색어가 급상승한다면?? 이진힙에서 최대힙으로 구현
        
        
4. upheap, downheap
    - 삽입할 때는 upheap() 이 사용된다. 즉, 노드 추가는 최하위에 삽입되고 해당 부모노드와 비교하여 작으면 더 위로 올라가게 되는 구조
    - 삭제할 때는 downheap() 이 사용된다. 최상위 노드를 삭제하고, 그 자리에 최하위 노드를 추가하여 둘이서 스와핑을 진행.
        - *스와핑: 최상위, 최하위 = 최하위, 최상위
        - 최하위가 완전부모 노드에 있기 때문에 자식 노드와 비교하면서 아래로 내려가게 되는 자료구조

In [24]:
class BHeap:
    def __init__(self, a): # 이진힙 생성자
        self.a = a # 리스트 a
        self.N = len(a) -1 # 항목 수
        
    def create_heap(self): # 초기힙 생성
        # start 는 항목 수, end 는 모든 항목이 순회할 때까지, step 는 역순으로
        # 즉, 항목 수 -1 해서 역순으로 이진힙을 순회
        for i in range(self.N//2, 0, -1):
            self.downheap(i)
            
    def insert(self, key_value):
        self.N += 1
        self.a.append(key_value)
        self.upheap(self.N)
        
    def delete_min(self): #최소값 삭제
        if self.N == 0:
            print('힙이 비어있음')
            return None
        # 최상위 노드에 있는 루트 노드가 최소값이 되며 이를 minimum 변수에 할당
        # minimum 변수에 해당 값을 할당했으니, a[1] 의 실제값은 레퍼런스 카운트 2
        # a[1] 이 다른 값을 참조해도 minimum 은 최소값을 참조한 상태이다(헷갈리지 말 것!!!!)
        minimum = a[1] 
        self.a[1], self.a[-1] = self.a[-1], self.a[1] # 최소값과 최대값을 스위칭
        del self.a[-1] # 스위칭 된 최소값을 삭제
        self.N -= 1 # 삭제했으니 항목 수도 -1
        self.downheap(1)
        return minimum
    
    def downheap(self, i):
        while 2 * i <= self.N:
            k = 2 * i
            # k 인덱스가 크다면 k+1 인덱스가 최소값으로 지정되어야 함
            if k < self.N and self.a[k][0] > self.a[k+1][0]:
                k += 1
            if self.a[i][0] < self.a[k][0]:
                break
            self.a[i], self.a[k] = self.a[k], self.a[i]
            i = k
            
    def upheap(self, j): # 힙 올라가며 힙속성 회복
        # j 는 항목수(리스트의 개수)
        while j > 1 and self.a[j//2][0] > self.a[j][0]:
            # 부모노드와 자식노드를 교환
            self.a[j], self.a[j//2] = self.a[j//2], self.a[j]
            j = j//2 # 노드가 한 층 올라감
            
    def print_heap(self):
        for i in range(1, self.N+1):
            print('[%2d' % self.a[i][0], self.a[i][1], ']', end='')
        print('\n힙 크기 = ', self.N)
                

In [25]:
a = [None] * 1
a.append([90, 'watermelon'])
a.append([80, 'pear'])
a.append([70, 'melon'])
a.append([60, 'lime'])
a.append([50, 'mango'])
a.append([40, 'cherry'])
a.append([30, 'grape'])
a.append([20, 'orange'])
a.append([10, 'apricot'])
a.append([15, 'banana'])
a.append([45, 'lemon'])
a.append([40, 'kiwi'])
b = BHeap(a)
print('힙 만들기 전:')
b.print_heap()
b.create_heap()
print('최소힙:')
b.print_heap()
print('최솟값 삭제 후')
print(b.delete_min())
b.print_heap()
b.insert([5,'apple'])
print('5 삽입 후')
b.print_heap()

힙 만들기 전:
[90 watermelon ][80 pear ][70 melon ][60 lime ][50 mango ][40 cherry ][30 grape ][20 orange ][10 apricot ][15 banana ][45 lemon ][40 kiwi ]
힙 크기 =  12
최소힙:
[10 apricot ][15 banana ][30 grape ][20 orange ][45 lemon ][40 kiwi ][70 melon ][80 pear ][60 lime ][50 mango ][90 watermelon ][40 cherry ]
힙 크기 =  12
최솟값 삭제 후
[10, 'apricot']
[15 banana ][20 orange ][30 grape ][40 cherry ][45 lemon ][40 kiwi ][70 melon ][80 pear ][60 lime ][50 mango ][90 watermelon ]
힙 크기 =  11
5 삽입 후
[ 5 apple ][20 orange ][15 banana ][40 cherry ][45 lemon ][30 grape ][70 melon ][80 pear ][60 lime ][50 mango ][90 watermelon ][40 kiwi ]
힙 크기 =  12


#### \# 2 차원 리스트

In [26]:
fruit = [[10, 'apple'], [20, 'banana'], [30, 'kiwi', 'steve']]

In [27]:
fruit

[[10, 'apple'], [20, 'banana'], [30, 'kiwi', 'steve']]

In [28]:
a = 2

In [29]:
fruit[a*1][0]

30

In [30]:
fruit[a*1][1]

'kiwi'

In [31]:
fruit[a*1][2] # 세번째 리스트에만 3개의 리스트값을 가지고 있어도 출력된다. 물론 다른 fruit 리스트들을 출력하면 에러 발생

'steve'

### \# 이진탐색

In [2]:
def binary_search(left, right, t):
    if left > right:
        return None # 탐색 실패(즉, t가 리스트에 없음)
    mid = (left + right) // 2 # 리스트에서 탐색할 부분의 중간 항목의 인덱스 계산
    if a[mid] == t:
        return mid # 탐색 성공
    if a[mid] > t:
        binary_search(left, mid-1, t) # 앞부분 탐색
    else:
        binary_search(mid+1, right, t) #뒷부분 탐색

1. 기본 전제는 a 리스트의 N 개 값들은 정렬이 되어있다.
2. mid 로 범위를 좁여나간다
3. 최악의 경우에는 리스트에 거의 대부분을 검색하여 return mid 가 된다.
    - 삽입과 삭제가 빈번하면 정렬을 유지하기 위해 시간이 오래 걸린다.
    - O(N): 선형시간(3번째)
    
    

[!TIP]
- binary_search(인덱스 첫번째, 인덱스 마지막, 원하는 검색값)
- binary_esarch(mid, 인덱스 마지막, 원하는 검색값)
- mid 를 +- 하면서 최종적으로는 binary_search(10, 10, 66) # 2개의 범위를 일치
- 조건에 만족하지 않으면 재귀함수로 호출하여 범위를 좁혀 나가는 것 또한 잊지말자.